### PR TITLE
Network game

### PR DESCRIPTION
**Pull request by ksdenisova**
**As a** player
**I want** to play a remote game
**So that** we can play together 
**And** be on different computers

**Given:** Tic-Tac-Toe 
**And** Choosing a network game
**When:** game is playing
**Then:** player one is a host
**And** player two joins to player one's game
**And** human vs human game is played over network

[Trello](https://trello.com/c/yPCQaPFp/30-network-game)

CleanShot 2021-11-11 at 16 47 57@2x

**Refactoring:**
- Refactor `InputOutput` - remove duplication code, created `choose_option` for the same behavioral.
- Add tests for `AbstractServer`, `AbstractClient`, `Server` and `Client`.
- Add exit option for the whole game.

### CODE REVIEW

File: lib/network/game_client.rb [hunk 1]
```diff
@@ -1,60 +1,39 @@
...
 
         def start
-            connect
-            communicate
-            close_connection
+            begin
+                connect
+                run
+            rescue Interrupt
+                @io.print_output(" Exiting now... Bye bye...")
```
Comment (Aoxian) [inline lib/network/game_client.rb:18]: this seems like an odd place for a message to be hiding out.

File: lib/network/game_server.rb [hunk 2]
```diff
@@ -16,11 +18,17 @@ def current_address
...
-            open_connection
-            print_connection_info
-            accept_connection
-            communicate
-            close_connection
+            begin
+                open_connection
+                print_connection_info
+                accept_connection
+                @io.print_output("Player one join the game")
```
Comment (Aoxian) [inline lib/network/game_server.rb:25]: here too; just "food for thought".

Comment (Aoxian) [reply lib/network/game_server.rb:25]: Proof reading comment; hope it is helpful.
```suggestion
                @io.print_output("Player one joined the game")
```

File: lib/tic_tac_toe.rb [hunk 3]
```diff
@@ -1,16 +1,20 @@
...
+        game.run
+    else
+        begin
+            setup.network_game.start
+        rescue StandardError
+            puts "Connection failed"
+        end
     end
+rescue Interrupt
+    puts " Exiting now... Bye bye..."
```
Comment (Aoxian) [inline lib/tic_tac_toe.rb:19]: this line of code is sprinkled everywhere.  I wonder if there is an object or a helper function missing.

File: spec/lib/game_engine_spec.rb [hunk 4]
```diff
@@ -61,7 +64,10 @@
 
             game = GameEngine.new(io_mock, human_player, computer_player_mock)
 
+            expect(io_mock).to receive(:print_welcome).once
```
Comment (Aoxian) [inline spec/lib/game_engine_spec.rb:67]: The name of this test is: `it 'when player chose unavailable place on board it waits for right move'`

These extreme mocking scenarios are smelly and hard to follow.  A simpler solution would create a board and then inject it into a game, saving most of the setup.  That way you can just test and prove the thing you are testing.

Comment (Aoxian) [reply spec/lib/game_engine_spec.rb:67]: Keep It Simple (KIS)

File: spec/lib/network/abstract_client_spec.rb [hunk 5]
```diff
@@ -0,0 +1,28 @@
+require 'network/abstract_client'
+
+describe Network::AbstractClient do
```
Comment (Aoxian) [inline spec/lib/network/abstract_client_spec.rb:3]: its probably OK to not test interfaces.  I know that sounds counter intuitive, and this is a dynamic language so its actually good that you did.  On a client I would consider this a nice to have.

File: spec/lib/remote_player_spec.rb [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+    end
+
+    context 'inizialize' do
+        it 'when a network game has chosen a remote player is initialized correctly' do
+            expect { RemotePlayer.new(@player_name, @player_mark) }.not_to raise_error
+        end
+
+        it 'returns true if RemotePlayer inherits Player' do
+            expect(RemotePlayer.ancestors.include?(Player)).to be true
+        end
```
Comment (Aoxian) [inline spec/lib/remote_player_spec.rb:24]: like testing interfaces, testing inheritance seems a like a nice to have. I am not sure I would do this in actual production code.

File: spec/lib/smart_computer_player_spec.rb [hunk 7]
```diff
@@ -32,7 +32,7 @@
             expect(@computer_player.make_move(@board)).to eq best_move
         end
 
-        it 'should return a differen move is a middle of the board is unavailable' do
+        it 'should return a differen move if a middle of the board is unavailable' do
```
Comment (Aoxian) [inline spec/lib/smart_computer_player_spec.rb:35]: This is an example of a good refactoring that is out of the scope of the ticket; it would probably be better served in a small "campsite" PR instead of trying to sneak it into an already very large PR.

**Review by Aoxian — APPROVED** [review]
This is a huge PR, looks like you put in a lot of work.  The larger the PR the harder it is for reviewers to review.  Overall I think it looks pretty good. The use hand spun mocks/spys for testing the server and client was a good move.  I am approving, and sprinkled in some comments.  Nothing is blocking, so do not feel like you need to respond to the comments in anyway at this time.

File: lib/network/game_server.rb (hunk 2 above)
Comment (ksdenisova) [reply lib/network/game_server.rb:25]: > This is a huge PR, looks like you put in a lot of work. The larger the PR the harder it is for reviewers to review. Overall I think it looks pretty good. The use of hand spun mocks/spys for testing the server and client was a good move. I am approving, and sprinkled in some comments. Nothing is blocking, so do not feel like you need to respond to the comments in anyway at this time.

I agree with all the comments you made in this PR. I'll follow your recommendations. Thank you very much!
